### PR TITLE
Extract function types from function and arrow expressions.

### PR DESCRIPTION
**Pull request by dragomirtitian**
This PR brings TS emit closer to what an external tool could emit without type information.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -6171,10 +6171,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const overrideTypeNode = typeViaParent && jsDocProperty.typeExpression && getTypeFromTypeNode(context, jsDocProperty.typeExpression.type) !== typeViaParent ? typeToTypeNodeHelper(typeViaParent, context) : undefined;
                 return overrideTypeNode;
             },
-            enterNewScope(context, node) {
+            enterNewScope(context, node, expandParameters = true) {
                 if (isFunctionLike(node) || isJSDocSignature(node)) {
                     const signature = getSignatureFromDeclaration(node);
-                    const expandedParams = getExpandedParameters(signature, /*skipUnionExpanding*/ true)[0];
+                    const expandedParams = !expandParameters ? signature.parameters : getExpandedParameters(signature, /*skipUnionExpanding*/ true)[0];
```
Comment (jakebailey) [inline src/compiler/checker.ts:6177]: Believe it or not, all tests pass if we just unconditionally use `signature.parameters` here (not expanding at all). I'm not an expert in why we needed to expand them before, but it sure feels like we don't need to do them at all anymore?

Comment (dragomirtitian) [reply src/compiler/checker.ts:6177]: You are right. Expansion no longer ever happens from the syntactic printer. Only in the type checker printer. Removed the option

**Comment by jakebailey** [comment]
@typescript-bot test it

File: tests/baselines/reference/declFileTypeofFunction.js [hunk 2]
```diff
@@ -70,6 +70,6 @@ declare function b1(): typeof b1;
 declare function foo(): typeof foo;
 declare var foo1: typeof foo;
 declare var foo2: typeof foo;
-declare var foo3: () => /*elided*/ any;
-declare var x: () => /*elided*/ any;
+declare var foo3: () => () => /*elided*/ any;
+declare var x: () => () => /*elided*/ any;
```
[the remark this answers is not in the document]
Comment (dragomirtitian) [reply tests/baselines/reference/declFileTypeofFunction.js:74]: The base cause is that at the root level syntactic printing is now always given a try. And in this case the parameter list can be copied, it's just the return type that needs to fallback on type printing. I think I can revert this behavior.

[the remark this answers is not in the document]
Comment (dragomirtitian) [reply tests/baselines/reference/declFileTypeofFunction.js:74]: I looked into it and I don't think I can easily revert to the old behavior. The type checked printer only knows a type is too recursive to represent when it tried to print it. So when I see the function expression, I try to reuse types from the expression. When doing so, I discover the missing return type and fallback to type checker printing which then discovers teh type is recursive. 

What specifically is the worry in this case ? `() => any` and `() => () => any` are both bad, they both allow calling the function, so I don't really think the types are worse.

[the remark this answers is not in the document]
Comment (jakebailey) [reply tests/baselines/reference/declFileTypeofFunction.js:74]: Moreso if someone actually calls it and uses the result, that result will not be `any` and may cause follow-on errors. But I think this particular case is exceeding rare and weird anyway. Like, the function is clearly returning a function, so maybe that safety is better.

**Review by jakebailey — APPROVED** [review]
In general, this seems fine to me, though I am unsure if we want to sneak this in before the 5.7 RC. @weswigham @DanielRosenwasser  do you have an opinion here?

**Comment by jakebailey** [comment]
5.7 is cut, main is open for 5.8; I'm just going to merge this now. I don't think we need this for 5.7 or anything.

**Pull request merged**